// @fullcalendar/core/internal.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var internalCommon = require('./internal-common.cjs');
require('preact');
require('preact/compat');



exports.BASE_OPTION_DEFAULTS = internalCommon.BASE_OPTION_DEFAULTS;
exports.BaseComponent = internalCommon.BaseComponent;
exports.BgEvent = internalCommon.BgEvent;
exports.CalendarImpl = internalCommon.CalendarImpl;
exports.CalendarRoot = internalCommon.CalendarRoot;
exports.ContentContainer = internalCommon.ContentContainer;
exports.CustomRenderingStore = internalCommon.CustomRenderingStore;
exports.DateComponent = internalCommon.DateComponent;
exports.DateEnv = internalCommon.DateEnv;
exports.DateProfileGenerator = internalCommon.DateProfileGenerator;
exports.DayCellContainer = internalCommon.DayCellContainer;
exports.DayHeader = internalCommon.DayHeader;
exports.DaySeriesModel = internalCommon.DaySeriesModel;
exports.DayTableModel = internalCommon.DayTableModel;
exports.DelayedRunner = internalCommon.DelayedRunner;
exports.ElementDragging = internalCommon.ElementDragging;
exports.ElementScrollController = internalCommon.ElementScrollController;
exports.Emitter = internalCommon.Emitter;
exports.EventContainer = internalCommon.EventContainer;
exports.EventImpl = internalCommon.EventImpl;
exports.Interaction = internalCommon.Interaction;
exports.MoreLinkContainer = internalCommon.MoreLinkContainer;
exports.NamedTimeZoneImpl = internalCommon.NamedTimeZoneImpl;
exports.NowIndicatorContainer = internalCommon.NowIndicatorContainer;
exports.NowTimer = internalCommon.NowTimer;
exports.PositionCache = internalCommon.PositionCache;
exports.RefMap = internalCommon.RefMap;
exports.ScrollController = internalCommon.ScrollController;
exports.ScrollResponder = internalCommon.ScrollResponder;
exports.Scroller = internalCommon.Scroller;
exports.SegHierarchy = internalCommon.SegHierarchy;
exports.SimpleScrollGrid = internalCommon.SimpleScrollGrid;
exports.Slicer = internalCommon.Slicer;
exports.Splitter = internalCommon.Splitter;
exports.StandardEvent = internalCommon.StandardEvent;
exports.TableDateCell = internalCommon.TableDateCell;
exports.TableDowCell = internalCommon.TableDowCell;
exports.Theme = internalCommon.Theme;
exports.ViewContainer = internalCommon.ViewContainer;
exports.ViewContextType = internalCommon.ViewContextType;
exports.WeekNumberContainer = internalCommon.WeekNumberContainer;
exports.WindowScrollController = internalCommon.WindowScrollController;
exports.addDays = internalCommon.addDays;
exports.addDurations = internalCommon.addDurations;
exports.addMs = internalCommon.addMs;
exports.addWeeks = internalCommon.addWeeks;
exports.allowContextMenu = internalCommon.allowContextMenu;
exports.allowSelection = internalCommon.allowSelection;
exports.applyMutationToEventStore = internalCommon.applyMutationToEventStore;
exports.applyStyle = internalCommon.applyStyle;
exports.asCleanDays = internalCommon.asCleanDays;
exports.asRoughMinutes = internalCommon.asRoughMinutes;
exports.asRoughMs = internalCommon.asRoughMs;
exports.asRoughSeconds = internalCommon.asRoughSeconds;
exports.binarySearch = internalCommon.binarySearch;
exports.buildElAttrs = internalCommon.buildElAttrs;
exports.buildEntryKey = internalCommon.buildEntryKey;
exports.buildEventApis = internalCommon.buildEventApis;
exports.buildEventRangeKey = internalCommon.buildEventRangeKey;
exports.buildIsoString = internalCommon.buildIsoString;
exports.buildNavLinkAttrs = internalCommon.buildNavLinkAttrs;
exports.buildSegTimeText = internalCommon.buildSegTimeText;
exports.collectFromHash = internalCommon.collectFromHash;
exports.combineEventUis = internalCommon.combineEventUis;
exports.compareByFieldSpecs = internalCommon.compareByFieldSpecs;
exports.compareNumbers = internalCommon.compareNumbers;
exports.compareObjs = internalCommon.compareObjs;
exports.computeEarliestSegStart = internalCommon.computeEarliestSegStart;
exports.computeEdges = internalCommon.computeEdges;
exports.computeFallbackHeaderFormat = internalCommon.computeFallbackHeaderFormat;
exports.computeInnerRect = internalCommon.computeInnerRect;
exports.computeRect = internalCommon.computeRect;
exports.computeShrinkWidth = internalCommon.computeShrinkWidth;
exports.computeVisibleDayRange = internalCommon.computeVisibleDayRange;
exports.config = internalCommon.config;
exports.constrainPoint = internalCommon.constrainPoint;
exports.createDuration = internalCommon.createDuration;
exports.createEmptyEventStore = internalCommon.createEmptyEventStore;
exports.createEventInstance = internalCommon.createEventInstance;
exports.createEventUi = internalCommon.createEventUi;
exports.createFormatter = internalCommon.createFormatter;
exports.diffDates = internalCommon.diffDates;
exports.diffDayAndTime = internalCommon.diffDayAndTime;
exports.diffDays = internalCommon.diffDays;
exports.diffPoints = internalCommon.diffPoints;
exports.diffWeeks = internalCommon.diffWeeks;
exports.diffWholeDays = internalCommon.diffWholeDays;
exports.diffWholeWeeks = internalCommon.diffWholeWeeks;
exports.disableCursor = internalCommon.disableCursor;
exports.elementClosest = internalCommon.elementClosest;
exports.elementMatches = internalCommon.elementMatches;
exports.enableCursor = internalCommon.enableCursor;
exports.eventTupleToStore = internalCommon.eventTupleToStore;
exports.filterHash = internalCommon.filterHash;
exports.findDirectChildren = internalCommon.findDirectChildren;
exports.findElements = internalCommon.findElements;
exports.flexibleCompare = internalCommon.flexibleCompare;
exports.formatDayString = internalCommon.formatDayString;
exports.formatIsoMonthStr = internalCommon.formatIsoMonthStr;
exports.formatIsoTimeString = internalCommon.formatIsoTimeString;
exports.getAllowYScrolling = internalCommon.getAllowYScrolling;
exports.getCanVGrowWithinCell = internalCommon.getCanVGrowWithinCell;
exports.getClippingParents = internalCommon.getClippingParents;
exports.getDateMeta = internalCommon.getDateMeta;
exports.getDayClassNames = internalCommon.getDayClassNames;
exports.getDefaultEventEnd = internalCommon.getDefaultEventEnd;
exports.getElSeg = internalCommon.getElSeg;
exports.getEntrySpanEnd = internalCommon.getEntrySpanEnd;
exports.getEventTargetViaRoot = internalCommon.getEventTargetViaRoot;
exports.getIsRtlScrollbarOnLeft = internalCommon.getIsRtlScrollbarOnLeft;
exports.getRectCenter = internalCommon.getRectCenter;
exports.getRelevantEvents = internalCommon.getRelevantEvents;
exports.getScrollGridClassNames = internalCommon.getScrollGridClassNames;
exports.getScrollbarWidths = internalCommon.getScrollbarWidths;
exports.getSectionClassNames = internalCommon.getSectionClassNames;
exports.getSectionHasLiquidHeight = internalCommon.getSectionHasLiquidHeight;
exports.getSegAnchorAttrs = internalCommon.getSegAnchorAttrs;
exports.getSegMeta = internalCommon.getSegMeta;
exports.getSlotClassNames = internalCommon.getSlotClassNames;
exports.getStickyFooterScrollbar = internalCommon.getStickyFooterScrollbar;
exports.getStickyHeaderDates = internalCommon.getStickyHeaderDates;
exports.getUniqueDomId = internalCommon.getUniqueDomId;
exports.greatestDurationDenominator = internalCommon.greatestDurationDenominator;
exports.groupIntersectingEntries = internalCommon.groupIntersectingEntries;
exports.guid = internalCommon.guid;
exports.hasBgRendering = internalCommon.hasBgRendering;
exports.hasCustomDayCellContent = internalCommon.hasCustomDayCellContent;
exports.hasShrinkWidth = internalCommon.hasShrinkWidth;
exports.identity = internalCommon.identity;
exports.injectStyles = internalCommon.injectStyles;
exports.interactionSettingsStore = internalCommon.interactionSettingsStore;
exports.interactionSettingsToStore = internalCommon.interactionSettingsToStore;
exports.intersectRanges = internalCommon.intersectRanges;
exports.intersectRects = internalCommon.intersectRects;
exports.intersectSpans = internalCommon.intersectSpans;
exports.isArraysEqual = internalCommon.isArraysEqual;
exports.isColPropsEqual = internalCommon.isColPropsEqual;
exports.isDateSelectionValid = internalCommon.isDateSelectionValid;
exports.isDateSpansEqual = internalCommon.isDateSpansEqual;
exports.isInt = internalCommon.isInt;
exports.isInteractionValid = internalCommon.isInteractionValid;
exports.isMultiDayRange = internalCommon.isMultiDayRange;
exports.isPropsEqual = internalCommon.isPropsEqual;
exports.isPropsValid = internalCommon.isPropsValid;
exports.isValidDate = internalCommon.isValidDate;
exports.mapHash = internalCommon.mapHash;
exports.memoize = internalCommon.memoize;
exports.memoizeArraylike = internalCommon.memoizeArraylike;
exports.memoizeHashlike = internalCommon.memoizeHashlike;
exports.memoizeObjArg = internalCommon.memoizeObjArg;
exports.mergeEventStores = internalCommon.mergeEventStores;
exports.multiplyDuration = internalCommon.multiplyDuration;
exports.padStart = internalCommon.padStart;
exports.parseBusinessHours = internalCommon.parseBusinessHours;
exports.parseClassNames = internalCommon.parseClassNames;
exports.parseDragMeta = internalCommon.parseDragMeta;
exports.parseEventDef = internalCommon.parseEventDef;
exports.parseFieldSpecs = internalCommon.parseFieldSpecs;
exports.parseMarker = internalCommon.parse;
exports.pointInsideRect = internalCommon.pointInsideRect;
exports.preventContextMenu = internalCommon.preventContextMenu;
exports.preventDefault = internalCommon.preventDefault;
exports.preventSelection = internalCommon.preventSelection;
exports.rangeContainsMarker = internalCommon.rangeContainsMarker;
exports.rangeContainsRange = internalCommon.rangeContainsRange;
exports.rangesEqual = internalCommon.rangesEqual;
exports.rangesIntersect = internalCommon.rangesIntersect;
exports.refineEventDef = internalCommon.refineEventDef;
exports.refineProps = internalCommon.refineProps;
exports.removeElement = internalCommon.removeElement;
exports.removeExact = internalCommon.removeExact;
exports.renderChunkContent = internalCommon.renderChunkContent;
exports.renderFill = internalCommon.renderFill;
exports.renderMicroColGroup = internalCommon.renderMicroColGroup;
exports.renderScrollShim = internalCommon.renderScrollShim;
exports.requestJson = internalCommon.requestJson;
exports.sanitizeShrinkWidth = internalCommon.sanitizeShrinkWidth;
exports.setRef = internalCommon.setRef;
exports.sliceEventStore = internalCommon.sliceEventStore;
exports.sortEventSegs = internalCommon.sortEventSegs;
exports.startOfDay = internalCommon.startOfDay;
exports.translateRect = internalCommon.translateRect;
exports.triggerDateSelect = internalCommon.triggerDateSelect;
exports.unpromisify = internalCommon.unpromisify;
exports.whenTransitionDone = internalCommon.whenTransitionDone;
exports.wholeDivideDurations = internalCommon.wholeDivideDurations;
